src/stdfblib/ita/OPCUA_MGR.cpp

```diff
@@ -949,6 +981,101 @@ UA_StatusCode OPCUA_MGR::onDeleteConnection(UA_Server*,
+  if (methodContext != nullptr) {
+    EMGMResponse eRetVal = EMGMResponse::UnsupportedType;
+    std::string destination;
+    destination = getInputValue(*static_cast<UA_String*>(input[0].data));
+    std::vector<std::string> fullFbName;
+    parseDestinationName(destination, fullFbName);
+
+    const char* resourceName = static_cast<const char*>(objectContext);
+    OPCUA_MGR* uaMGR = static_cast<OPCUA_MGR*>(methodContext);
+    uaMGR->setMGMCommand(EMGMCommandType::MonitoringAddWatch, CStringDictionary::getInstance().insert(resourceName), nullptr, fullFbName);
+    eRetVal = uaMGR->mUaDevice.executeMGMCommand(uaMGR->mCommand);
+    return scResponseMap.find(eRetVal)->second;
+  }
+  return UA_STATUSCODE_BADUNKNOWNRESPONSE;
+}
+
+EMGMResponse OPCUA_MGR::addReadWatchesMethod(UA_Server* paServer) {
+  UA_Argument outputArgument;
+  initArgument(outputArgument, UA_TYPES_STRING, smReadWatchesOutArgName, smReadWatchesOutArgDescription);
+  UA_MethodAttributes readWatchesAttr = createAttribute(smReadWatchesDisplayName, smReadWatchesDescription);
+  return addMethodNode(paServer, smReadWatchesMethodName, mMgmtTypeId, readWatchesAttr, nullptr, 0, &outputArgument, 1, &onReadWatches);
+}
+
+UA_StatusCode OPCUA_MGR::onReadWatches(UA_Server*,
+  const UA_NodeId*, void*,
+  const UA_NodeId*, void* methodContext,
+  const UA_NodeId*, void*,
+  size_t, const UA_Variant*,
+  size_t, UA_Variant* output) {
+  if (methodContext != nullptr) {
+    EMGMResponse eRetVal = EMGMResponse::UnsupportedType;
+    OPCUA_MGR* uaMGR = static_cast<OPCUA_MGR*>(methodContext);
+    uaMGR->setMGMCommand(EMGMCommandType::MonitoringReadWatches, CStringDictionary::scmInvalidStringId, nullptr, nullptr, nullptr);
+    eRetVal = uaMGR->mUaDevice.executeMGMCommand(uaMGR->mCommand);
+    int status = scResponseMap.find(eRetVal)->second;
+    if (status != UA_STATUSCODE_GOOD) {
+      return status;
+    }
+    UA_String uaResp = UA_String_fromChars(uaMGR->mCommand.mMonitorResponse.c_str());
+    status = UA_Variant_setScalarCopy(output, &uaResp, &UA_TYPES[UA_TYPES_STRING]);
+    UA_String_clear(&uaResp);
+    return status;
+    }
+  return UA_STATUSCODE_BADUNKNOWNRESPONSE;
+}
+
+EMGMResponse OPCUA_MGR::addRemoveWatchMethod(UA_Server* paServer) {
+  UA_Argument inputArgument;
+  initArgument(inputArgument, UA_TYPES_STRING, smRemoveWatchArgName, smRemoveWatchArgDescription);
+
+  UA_MethodAttributes removeWatchAttr = createAttribute(smRemoveWatchAttrDisplayName, smRemoveWatchAttrDescription);
+  return addMethodNode(paServer, smRemoveWatchMethodName, mResourceTypeId, removeWatchAttr, &inputArgument, 1, nullptr, 0, &onRemoveWatch);
+}
+
+UA_StatusCode OPCUA_MGR::onRemoveWatch(UA_Server*,
+  const UA_NodeId*, void*,
+  const UA_NodeId*, void* methodContext,
+  const UA_NodeId*, void* objectContext,
+  size_t, const UA_Variant* input,
+  size_t, UA_Variant*) {
+  if (methodContext != nullptr) {
+    EMGMResponse eRetVal = EMGMResponse::UnsupportedType;
+    std::string destination;
```

Comment:
Could these two lines be combined into one?
`std::string destination(getInputValue(*static_cast<UA_String*>(input[0].data));)` or
`std::string destination = getInputValue(*static_cast<UA_String*>(input[0].data));`

By passing the value to the ctor, the string storage would be initialized in the appropriate size directly, and you could make `std::string destination` even `const`.

While outside this review, I would also recommend making the first parameter of the `parseDestinationName` also `const`, as the string is never modified

Reply:
Fixed in new commit